File: hdf-parsers/typescript/saf-supplement.ts
// Backwards-compatible normalization of legacy SAF-supplement top-level keys
// (`target`, `passthrough`) into v3-native carriers, run before schema
// validation. Mirrors the Go side at hdf-parsers/go/saf_supplement.go — the two
// are pinned against a shared legacy-in/v3-out fixture pair.

const SAF_TARGET_DEPRECATION =
  "deprecated: top-level 'target' (saf supplement target write) was normalized into components[]; emit a component directly — this compatibility shim will be removed in a future release";
const SAF_PROVENANCE_DEPRECATION =
  "deprecated: top-level 'passthrough' was normalized into extensions.passthrough; write provenance under extensions — this compatibility shim will be removed in a future release";

// v3 component/target types. Kept in sync with the schema's Target_Type enum; an
// unknown type is left for the schema to reject rather than guessed.
const VALID_COMPONENT_TYPES = new Set([
  'aiModel', 'application', 'artifact', 'cloudAccount', 'cloudResource',
  'containerImage', 'containerInstance', 'containerPlatform', 'database',
  'dataset', 'host', 'network', 'repository',
]);

export interface SafNormalizeResult {
  /** The normalized JSON string (byte-identical to the input when nothing was rewritten). */
  output: string;
  /** One deprecation warning per rewritten legacy key. */
  warnings: string[];
}

function isObject(v: unknown): v is Record<string, unknown> {
  return typeof v === 'object' && v !== null && !Array.isArray(v);
}

/**
 * Rewrite the legacy SAF-supplement top-level keys into v3-native carriers so a
 * SAF-produced document becomes valid v3 before schema validation: `target` → a
 * components[] entry (type validated; id → name and, for cloudAccount, accountId;
 * boundary → labels.boundary; merged into a matching existing component rather
 * than duplicated); `passthrough` → extensions.passthrough.
 *
 * Deliberately narrow: input that is not a JSON object, or carries neither legacy
 * key, is returned BYTE-IDENTICAL with no warnings. A target whose type is not a
 * valid component type is left in place (with a warning) for the schema to reject.
 */
export function normalizeSafSupplement(input: string): SafNormalizeResult {
  let doc: unknown;
  try {
    doc = JSON.parse(input);
  } catch {
    return { output: input, warnings: [] }; // not JSON — let the validator surface it
  }
  if (!isObject(doc) || (!('target' in doc) && !('passthrough' in doc))) {
    return { output: input, warnings: [] }; // no legacy keys — byte-identical
  }

  const warnings: string[] = [];

  if ('target' in doc) {
    if (rewriteTarget(doc)) {
      warnings.push(SAF_TARGET_DEPRECATION);
    } else {
      warnings.push(
        `${SAF_TARGET_DEPRECATION} (target.type is not a recognized component type; left for schema validation)`,
      );
    }
  }

  if ('passthrough' in doc) {
    rewritePassthrough(doc);
    warnings.push(SAF_PROVENANCE_DEPRECATION);
  }

  return { output: JSON.stringify(doc), warnings };
}

/** Lift a legacy top-level `target` object into components[]; returns false
 * (leaving `target` in place) when it is not an object or its type is invalid. */
function rewriteTarget(doc: Record<string, unknown>): boolean {
  const tgt = doc.target;
  if (!isObject(tgt)) return false;
  const typ = typeof tgt.type === 'string' ? tgt.type : '';
  if (!VALID_COMPONENT_TYPES.has(typ)) return false;
  const id = typeof tgt.id === 'string' ? tgt.id : '';
  const boundary = typeof tgt.boundary === 'string' ? tgt.boundary : '';

  const comps: unknown[] = Array.isArray(doc.components) ? doc.components : [];

  let merged = false;
  for (const c of comps) {
    if (isObject(c) && c.name === id && c.type === typ) {
      applyBoundaryLabel(c, boundary);
      merged = true;
      break;
    }
  }
  if (!merged) {
    const c: Record<string, unknown> = { type: typ, name: id };
    if (typ === 'cloudAccount') c.accountId = id;
    applyBoundaryLabel(c, boundary);
    comps.push(c);
  }
  doc.components = comps;
  delete doc.target;
  return true;
}

/** Record the legacy target.boundary as a component label (the deferred
 * first-class-field decision — ADR-0015). No-op for an empty boundary. */
function applyBoundaryLabel(c: Record<string, unknown>, boundary: string): void {
  if (boundary === '') return;
  const labels = isObject(c.labels) ? c.labels : {};
  if (!('boundary' in labels)) labels.boundary = boundary;
  c.labels = labels;
}

/** Move the legacy top-level `passthrough` under extensions.passthrough, merging
 * into any existing extensions without clobbering other keys. */
function rewritePassthrough(doc: Record<string, unknown>): void {
  const ext = isObject(doc.extensions) ? doc.extensions : {};
  if (!('passthrough' in ext)) ext.passthrough = doc.passthrough;
  doc.extensions = ext;
  delete doc.passthrough;
}

File: hdf-parsers/go/saf_supplement.go
package hdfparsers

import (
	"encoding/json"
	"fmt"

	hdf "github.com/mitre/hdf-libs/hdf-schema/dist/go/v3"
)

// safTargetDeprecation and safProvenanceDeprecation name the v3-native shape so
// users migrate off the legacy SAF-supplement convention.
const (
	safTargetDeprecation     = "deprecated: top-level 'target' (saf supplement target write) was normalized into components[]; emit a component directly — this compatibility shim will be removed in a future release"
	safProvenanceDeprecation = "deprecated: top-level 'passthrough' was normalized into extensions.passthrough; write provenance under extensions — this compatibility shim will be removed in a future release"
)

// validComponentType reports whether s is a v3 component/target type. Kept in
// sync with the TargetType enum; an unknown type is left for the schema to
// reject rather than guessed into a wrong shape.
func validComponentType(s string) bool {
	switch hdf.TargetType(s) {
	case hdf.AIModel, hdf.Application, hdf.Artifact, hdf.CloudAccount,
		hdf.CloudResource, hdf.ContainerImage, hdf.ContainerInstance,
		hdf.ContainerPlatform, hdf.Database, hdf.Dataset, hdf.Host,
		hdf.Network, hdf.Repository:
		return true
	}
	return false
}

// NormalizeSAFSupplement rewrites the legacy SAF-supplement top-level keys into
// v3-native carriers so a SAF-produced document becomes valid v3 before schema
// validation: `target` → a components[] entry (type validated; id → name and,
// for cloudAccount, accountId; boundary → labels.boundary; merged into a
// matching existing component rather than duplicated); `passthrough` →
// extensions.passthrough. It returns the (possibly rewritten) document plus a
// deprecation warning per rewritten key.
//
// It is deliberately narrow: input that is not JSON, or carries neither legacy
// key, is returned BYTE-IDENTICAL with no warnings. A target whose type is not a
// valid component type is left in place (with a warning) for the schema to
// reject — never guessed. Mirrors hdf-parsers/typescript normalizeSafSupplement.
func NormalizeSAFSupplement(input []byte) ([]byte, []string) {
	var doc map[string]any
	if err := json.Unmarshal(input, &doc); err != nil {
		return input, nil // not a JSON object — let the validator surface it
	}
	_, hasTarget := doc["target"]
	_, hasPassthrough := doc["passthrough"]
	if !hasTarget && !hasPassthrough {
		return input, nil // no legacy keys — byte-identical passthrough
	}

	var warnings []string

	if hasTarget {
		if rewritten := rewriteTarget(doc); rewritten {
			warnings = append(warnings, safTargetDeprecation)
		} else {
			warnings = append(warnings, fmt.Sprintf(
				"%s (target.type is not a recognized component type; left for schema validation)", safTargetDeprecation))
		}
	}

	if hasPassthrough {
		rewritePassthrough(doc)
		warnings = append(warnings, safProvenanceDeprecation)
	}

	out, err := json.Marshal(doc)
	if err != nil {
		return input, nil // unreachable for a map decoded from valid JSON
	}
	return out, warnings
}

// rewriteTarget lifts a legacy top-level `target` object into components[].
// Returns false (leaving `target` in place) when the target is not an object or
// its type is not a valid component type.
func rewriteTarget(doc map[string]any) bool {
	tgt, ok := doc["target"].(map[string]any)
	if !ok {
		return false
	}
	typ, _ := tgt["type"].(string)
	if !validComponentType(typ) {
		return false
	}
	id, _ := tgt["id"].(string)
	boundary, _ := tgt["boundary"].(string)

	comps, _ := doc["components"].([]any)

	// Merge into an existing component with the same name+type rather than
	// duplicating; otherwise append a new one.
	merged := false
	for i := range comps {
		c, ok := comps[i].(map[string]any)
		if !ok {
			continue
		}
		if name, _ := c["name"].(string); name == id {
			if ct, _ := c["type"].(string); ct == typ {
				applyBoundaryLabel(c, boundary)
				merged = true
				break
			}
		}
	}
	if !merged {
		c := map[string]any{"type": typ, "name": id}
		if typ == string(hdf.CloudAccount) {
			c["accountId"] = id
		}
		applyBoundaryLabel(c, boundary)
		comps = append(comps, c)
	}
	doc["components"] = comps
	delete(doc, "target")
	return true
}

// applyBoundaryLabel records the legacy target.boundary as a component label
// (the deferred first-class-field decision — ADR-0015). No-op for an empty
// boundary.
func applyBoundaryLabel(c map[string]any, boundary string) {
	if boundary == "" {
		return
	}
	labels, ok := c["labels"].(map[string]any)
	if !ok {
		labels = map[string]any{}
	}
	if _, exists := labels["boundary"]; !exists {
		labels["boundary"] = boundary
	}
	c["labels"] = labels
}

// rewritePassthrough moves the legacy top-level `passthrough` under
// extensions.passthrough, merging into any existing extensions without
// clobbering other keys.
func rewritePassthrough(doc map[string]any) {
	ext, ok := doc["extensions"].(map[string]any)
	if !ok {
		ext = map[string]any{}
	}
	if _, exists := ext["passthrough"]; !exists {
		ext["passthrough"] = doc["passthrough"]
	}
	doc["extensions"] = ext
	delete(doc, "passthrough")
}
